### DISCUSSION TITLE
Callbacks as Rails

### DISCUSSION THREAD

**Discussion by abournier** [groue/GRDB.swift#1073] [Ideas]
Hello,

I just start learning GRDB. It looks like amazing.
However, is there a feature like Callbacks in Rails ? (beforeSave, afterSave, beforeCreate...)

I have found "didInsert" after an insert is performed. But is there an equivalent to "beforeSave" ? It could be useful, for example, to share code between multiple models. 

As an example, I used this to generate at the right place an Id as UUID string: 

```
extension Wine: MutablePersistableRecord {

		// MutablePersistableRecord protocol
		mutating func insert(_ db: Database) throws {
				generateUuidStringAsId()
				do {
						try performInsert(db)
				}
				catch {
						print("Unable to write the record !")
				}
		}
		
}
```

But I have 10 models using an UUID string as Id. If I could have something like:

```
beforeCreate() {
  generateUuidStringAsId()
} 
```

without having to implement `insert()` for all my models, it could be very helpful.

**Comment by groue** [comment]
Hello @abournier,

You shouldn't catch the insertion error. I don't understand why so many developers have such a urge to hide errors by catching and printing them, so that nobody can never look at it and handle it once the error happens in production. It's important to be able to look at errors straight in the eyes, because _this_ is the real way to make applications robust.

Your sample code should instead read as below, with all the useless error handling removed:

```swift
extension Wine {
    func insert(_ db: Database) throws {
        generateUuidStringAsId()
        try performInsert(db)
    }
}
```

But now I don't find that it's that much heavier than:

```swift
extension Wine {
    func beforeCreate() {
        generateUuidStringAsId()
    }
}
```

And I don't understand how callbacks would make your code _significantly_ simpler. I understand you may be used to ActiveRecord callbacks, but I'm not sure it's significantly more difficult to get used to the GRDB technique.

**Reply by groue** [reply]
I mean that your initial argument in favor of callbacks does not convince me. But I'm keen to change my mind! Please just provide stronger arguments.

**Reply by groue** [reply]
An argument I could listen to would be:

Considering ActiveRecord has been such an inspiration for GRDB, why not adding support for [callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks)?

It would help GRDB users who have prior experience with ActiveRecord profit from a familiar API. Since ActiveRecord callbacks are very popular, chances are other users would be happy using them as well.

Callbacks would probably be renamed in order to fit the platform, which prefers `will`/`did` over `before`/`after`: `willSave`, `didUpdate`, etc.

---

A counter argument I would also listen to is:

ActiveRecord callbacks are rooted in ActiveRecord, ActiveSupport, and generally Ruby classes and modules. In RoR, a record type can define its own callbacks, but it can also inherit callbacks from its super class, and from included modules or concerns. See for example [this question on Stack Overflow](https://stackoverflow.com/questions/7444522/is-it-possible-to-define-a-before-save-callback-in-a-module) which has a lot of up-votes, showing an actual interest of the RoR community in defining ActiveRecord callbacks in modules.

On the other hand, GRDB record types are often plain structs, and there is no runtime equivalent, in Swift, to Ruby's `include`, or ActiveSupport Concerns. All in all, callbacks in GRDB would always feel limited and crippled when compared to their Ruby inspiration, and they may not be worth the pain.

---

Another counter argument:

Callbacks may create a performance toll on all records, even records that do not define callbacks.

One behavior of ActiveRecord callbacks is very particular: a class inherits callbacks from all its super-classes, without having to call `super`.

Such behavior requires [external storage](https://github.com/rails/rails/blob/720f6b1ec21fb112522e664f86dfc28c5edbaefa/activesupport/lib/active_support/callbacks.rb#L508) for the definition of callbacks, and the iteration of registered callbacks on every call of a persistence method. The Swift compiler would not be able to optimize the case of types that do not define any callbacks. There lies the probable performance toll (to be measured).

We may prefer to use a `super`-based definition of callbacks, with regular overridable methods, in order to avoid the problems described in the above paragraph. But then RoR users may be surprised that callbacks are not automatically inherited.

---

Another counter argument:

ActiveRecord calls `before_destroy` and `after_destroy` on records, including associated records which are destroyed with their parent. GRDB can't do that. GRDB does not fetch records before deleting them, so it can't call the `destroy` callbacks. GRDB relies on foreign keys defined in the SQLite schema (`DELETE CASCADE`) for deleting dependent database rows, when ActiveRecord relies on `dependent: :destroy`, declared in Ruby code. (In practice, `before_destroy` is [quite subtle](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Ordering+callbacks)).

RoR users may be surprised that GRDB callbacks do not perform in the same way as they do in ActiveRecord.

---

This shows you the kind of input I'm looking for: arguments, counter-arguments, answers to arguments, and a will to make a library that serves well all its users.

**Reply by abournier** [reply]
> Hello @abournier,
> 
> You shouldn't catch the insertion error. I don't understand why so many developers have such a urge to hide errors by catching and printing them, so that nobody can never look at it and handle it once the error happens in production. It's important to be able to look at errors straight in the eyes, because _this_ is the real way to make applications robust.
> 
> Your sample code should instead read as below, with all the useless error handling removed:
> 
> ```swift
> extension Wine {
>     func insert(_ db: Database) throws {
>         generateUuidStringAsId()
>         try performInsert(db)
>     }
> }
> ```

Thanks for the suggestion. It's working (just need mutating keyword in my case)
I think some developers like me, are influenced by the debugger and add `do...catch` too much to quiet the debugger.

> 
> But now I don't find that it's that much heavier than:
> 
> ```swift
> extension Wine {
>     func beforeCreate() {
>         generateUuidStringAsId()
>     }
> }
> ```
> 
> And I don't understand how callbacks would make your code _significantly_ simpler. I understand you may be used to ActiveRecord callbacks, but I'm not sure it's significantly more difficult to get used to the GRDB technique.

In Rails, `beforeSave()` is called by AR automatically when saving the record. In your example, I need to call myself `beforeCreate()` somewhere, so I suppose in the insert() method too..

> An argument I could listen to would be:
> 
> Considering ActiveRecord has been such an inspiration for GRDB, why not adding support for [callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks)?
> 
> It would help GRDB users who have prior experience with ActiveRecord profit from a familiar API. Since ActiveRecord callbacks are very popular, chances are other users would be happy using them as well.
> 
> Callbacks would probably be renamed in order to fit the platform, which prefers `will`/`did` over `before`/`after`: `willSave`, `didUpdate`, etc.

Thanks for the writing suggestion. Yes, It's what I have in mind.
I read your counter arguments, but I'm not as expert as you in Rails internal mechanisms. Callbacks were just a suggestion, after seeing the way you build GRDB associations as Rails (has_many, belongs_to...).

**Reply by groue** [reply]
> In Rails, `beforeSave()` is called by AR automatically when saving the record. In your example, I need to call myself beforeCreate() somewhere, so I suppose in the `insert()` method too...

Yes @abournier. I guess it would look like, according to the [ActiveRecord callback ordering](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks):

```swift
extension Wine {
    func beforeSave() { ... }
    func beforeCreate() { ... }
    func beforeUpdate() { ... }

    func afterCreate() { ... }
    func afterSave() { ... }
    func afterUpdate() { ... }

    mutating func insert(_ db: Database) throws {
        beforeSave()
        beforeCreate()
        try performInsert(db)
        afterCreate()
        afterSave()
    }

    func update(...) throws {
        beforeSave()
        beforeUpdate()
        try performUpdate(...)
        afterUpdate()
        afterSave()
    }
}
```

You will have to override `save` as well, because by default this method [attempts at updating before inserting](https://github.com/groue/GRDB.swift/blob/32b2923e890df320906e64cbd0faca22a8bfda14/GRDB/Record/PersistableRecord.swift#L420-L432). This may trigger "before update" callbacks _and then_ "before create" callbacks. This is left as an exercice for the reader.

**Reply by groue** [reply]
> I read your counter arguments, but I'm not as expert as you in Rails internal mechanisms.

When I am asked to implement such a feature, I have to wonder, as anybody, "but how would I do that??". Eventually I have to learn things. You can do it as well, @abournier.

All my arguments and counter-arguments above are thought experiments where the only rule is to _take your request seriously_. From that I try to derive consequences, so that I can evaluate the request. You can do it as well.

> Callbacks were just a suggestion, after seeing the way you build GRDB associations as Rails (has_many, belongs_to...).

I 100% understand, and I'm happy this inspiration has helped you entering GRDB :-)

**Reply by abournier** [reply]
> Yes @abournier. I guess it would look like, according to the [ActiveRecord callback ordering](https://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks):
> 
> ```swift
> extension Wine {
>     func beforeSave() { ... }
>     func beforeCreate() { ... }
>     func beforeUpdate() { ... }
> 
>     func afterCreate() { ... }
>     func afterSave() { ... }
>     func afterUpdate() { ... }
> 
>     mutating func insert(_ db: Database) throws {
>         beforeSave()
>         beforeCreate()
>         try performInsert(db)
>         afterCreate()
>         afterSave()
>     }
> 
>     func update(...) throws {
>         beforeSave()
>         beforeUpdate()
>         try performUpdate(...)
>         afterUpdate()
>         afterSave()
>     }
> }
> ```
> 
> You will have to override `save` as well, because by default this method [attempts at updating before inserting](https://github.com/groue/GRDB.swift/blob/32b2923e890df320906e64cbd0faca22a8bfda14/GRDB/Record/PersistableRecord.swift#L420-L432). This may trigger "before update" callbacks _and then_ "before create" callbacks. This is left as an exercice for the reader.

Thanks for the code. The callbacks positions seem logical.

> > I read your counter arguments, but I'm not as expert as you in Rails internal mechanisms.
> 
> When I am asked to implement such a feature, I have to wonder, as anybody, "but how would I do that??". Eventually I have to learn things. You can do it as well, @abournier.
> 
> All my arguments and counter-arguments above are thought experiments where the only rule is to _take your request seriously_. From that I try to derive consequences, so that I can evaluate the request. You can do it as well.

Before asking, I had a look at the entire MutablePersistenceRecord protocol to see how things were made with `insert()`, `didInsert()` and the default implementation methods. My first idea was to extend this protocol for my needs.

Well, at the time, my only need was to have a way to not repeat the uuid creation for each model. That's why more generally why the callback idea come to me.
I also created another protocol `UuidStringAsId` that creates the uuid as string, with a default implementation, that my models can conform to.

Thanks for the proposals,
Have a good week-end,

**Reply by groue** [reply]
> Well, at the time, my only need was to have a way to not repeat the uuid creation for each model. That's why more generally why the callback idea come to me. I also created another protocol `UuidStringAsId` that creates the uuid as string, with a default implementation, that my models can conform to.

You wanted several records to perform a custom insertion (generate a UUID) simply by inheriting a protocol, right? Maybe you had something like the sample code below in mind?

```swift
protocol UuidStringAsId: MutatingPersistableRecord {
    var uuid: UUID? { get set }
}

// Conforming to UuidStringAsId grants automatic UUID generation on insert
extension UuidStringAsId {
    // Default implementation for the callback.
    mutating func beforeCreate() {
        if uuid == nil { uuid = UUID() }
    }
}

extension Wine: UuidStringAsId { }
extension Grape: UuidStringAsId { }
extension etc: UuidStringAsId { }
```

My concern with such inspiration from ActiveSupport Concerns (pun intended) is that with such a setup, you quickly hit a wall. Once a concern has been defined, it is natural to want more. For example:

```swift
protocol CreationDateTracking {
    var creationDate: Date? { get set }
}

extension CreationDateTracking {
    mutating func beforeCreate() {
        creationDate = Date()
    }
}

extension Wine: UuidStringAsId, CreationDateTracking { }
extension Grape: UuidStringAsId, CreationDateTracking { }
extension etc: UuidStringAsId, CreationDateTracking { }
```

I guess you would expect both callbacks to be run, right? Unfortunately, the Swift compiler disagrees.

My role here is to make sure we do not *over-promise*. Opening a door (hey, look at Rails-like callbacks!) and closing it too soon due to a bad match between a foreign feature and the language is a clear disservice to the library and its users.

Your feature request _is relevant_. I just don't know how to implement it today, in a way that supports all the wonderful variety of life (the wild needs of other users).

**Reply by abournier** [reply]
Completely true. So I forgot the callbacks,  and just used this approach at the time:

```
protocol UuidStringAsId {
		
		var id: String? { set get }
		
		mutating func generateUuidStringAsId()
}

extension UuidStringAsId {

		mutating func generateUuidStringAsId() {
				if id == nil {
						id = UUID().uuidString.lowercased()
				}
		}		
}

struct Wine: Codable, Identifiable, UuidStringAsId  {
		
		// Database table name
		static let databaseTableName = "usr_wines"
		
		// Properties
		var id: String?		
		var name: String
		var countryId: Int64?
}

// Model persistence
extension Wine: MutablePersistableRecord {

		// MutablePersistableRecord protocol
		mutating func insert(_ db: Database) throws {
				generateUuidStringAsId()
				try performInsert(db)
		}		
}
```

If  I need new protocols, I can create them and add them in insert() method.

**Comment by groue** [comment]
@abournier, GRDB 6 has [callbacks](https://github.com/groue/GRDB.swift/blob/master/README.md#persistence-callbacks) :-)

**Reply by abournier** [reply]
Hello @groue,

It looks amazing.  I will test them as soon as possible.
Thanks for all your energy given to this project.

**Reply by groue** [reply]
You're welcome :-) I pretty much copied the Rails callbacks: `willInsert` matches `before_create`, etc.
